### DISCUSSION TITLE
Long times of CopyHostToDevice

### DISCUSSION THREAD

**Discussion by CrimsonDump** [ROCm/ROCm#3134] [Q&A]
I've been running llama 2 with vllm, and trace with rocprof. This is what I got,  ~8 ms CopyHostToDevice CUDA time in a ~ 12 ms deocode phase, though these copies seem to be harmless to the core kernels (paged_attention, gemm, etc.. ). Then I found some clues from tracing which suggest that these CopyHostToDevice is for computing logits in Sampler, which confuses me because I thought all Weight were loaded at very beginning (nv did), instead of each decode phase.

So I'm wondering what are these CopyHostToDevice doing? Is it true sampler need to load weight every decode phase? 

<img width="762" alt="detail" src="https://github.com/ROCm/ROCm/assets/56749892/d14d7cbe-64f3-4a7e-bb3b-a51b1c9f68ad">
<img width="580" alt="one_decode" src="https://github.com/ROCm/ROCm/assets/56749892/4a738e84-2586-47eb-bea3-dead97f95070">
<img width="950" alt="detail_args" src="https://github.com/ROCm/ROCm/assets/56749892/3e9f77e3-0ae5-4aa2-a8a6-3a1a8da2fd79">
<img width="580" alt="code" src="https://github.com/ROCm/ROCm/assets/56749892/b9d6eb29-cacb-496d-9991-e27855d3eab6">

**Comment by sohaibnd** [comment]
Hi @CrimsonDump, sorry for the delay. Can you provide a sample of the llama 2 workload you are running with vllm as well as how you are generating the trace.

**Comment by sohaibnd** [comment]
I set up vllm (following [these instructions](https://rocm.blogs.amd.com/artificial-intelligence/vllm/README.html)) and traced a single prompt to LLama 2 (meta-llama/Llama-2-7b-chat-hf) using rocprof. I got the following trace:

![image](https://github.com/user-attachments/assets/59917eee-8dec-4591-94d2-a584e2fb8cd5)

I observed two places with host to device copies (labeled above):

1) These are in the order of milliseconds but they are at the start so it is probably just copying the model weights over to the gpu.

2) These are towards the end of the trace but the slices are much shorter, they are in the order of microseconds **not milliseconds**.

@CrimsonDump I'm going to close the issue for now since I am not able to reproduce these long CopyHostToDevice times but if you are still observing them, please provide more details on your workload so that I can reproduce the trace to investigate.
